### PR TITLE
Knative function controller component

### PR DESCRIPTION
**Pull request by venturasr**
Migrate runtime function controller from incubator into kyma components repo

Related issue(s)
kyma-project/kyma#4803

Related Epic
kyma-project/console#804

### CODE REVIEW

File: components/knative-function-controller/CODEOWNERS [hunk 1]
```diff
@@ -0,0 +1 @@
+/ @rakesh-garimella @sayanh @venturasr @lilitgh @k15r @nachtmaar
```
Comment (anishj0shi) [inline components/knative-function-controller/CODEOWNERS]: Do you need a separate CODEOWNERS for this component?, i think it should be covered via CODEOWNERS file at root level

File: components/knative-function-controller/.gitignore [hunk 2]
```diff
@@ -0,0 +1,25 @@
+
```
Comment (anishj0shi) [inline components/knative-function-controller/.gitignore]: shouldn't we enter regex in the .gitignore at root level?
something like **/knative-function-controller/*.exe,  **/knative-function-controller/*.out, etc

File: components/knative-function-controller/PROJECT [hunk 3]
```diff
@@ -0,0 +1,3 @@
+version: "1"
+domain: kyma-project.io
+repo: github.com/kyma-incubator/runtime
```
Comment (anishj0shi) [inline components/knative-function-controller/PROJECT]: is it still part of incubator?

File: components/knative-function-controller/README.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+Run the controller on your machine:
+
+```bash
+make run
+```
+
+#### Manager running inside k8s cluster
+
+This workflow needs to be used until [Come up with webhook developer workflow to test it locally #400](https://github.com/kubernetes-sigs/kubebuilder/issues/400) is fixed.
```
Comment (antoineco) [inline components/knative-function-controller/README.md]: tip: on GitHub issue references can be shortened like kubernetes-sigs/kubebuilder#40 (which also works as is inside markdown files!)

File: components/knative-function-controller/README.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+```
+
+#### Prod Deployment
+
+Uncomment `manager_image_patch_dev` in `kustomization.yaml`
+Then run the following commands:
+
+```bash
+make install
+APP_NAME = knative-function-controller
```
Comment (antoineco) [inline components/knative-function-controller/README.md]: No spaces before/after "=" in Bash (`bash: APP_NAME: command not found`)

File: components/knative-function-controller/README.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
+# Knative Function Controller
+
+Modify the project title and insert the name of your project. Use Heading 1 (H1).
+
+## Overview
+
+The Knative Function Controller is a Kubernetes Controller that enable Kyma to manage Function resources.
+
+It defines and handles a function Custom Resource Definition with the help of Knative Build and Knative Serving. Basically it is the serverless implementation based on knative.
```
Comment (antoineco) [inline components/knative-function-controller/README.md]: typos
function -> Function
knative -> Knative

File: components/knative-function-controller/README.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+make deploy
+```
+
+### Test
+
+```bash
+make test
+```
+
+#### Examples
```
Comment (antoineco) [inline components/knative-function-controller/README.md]: Where are they?

File: components/knative-function-controller/CODEOWNERS (hunk 1 above)
Comment (venturasr) [reply components/knative-function-controller/CODEOWNERS]: True we have CODEOWNERS file at root repo.

File: components/knative-function-controller/README.md (hunk 7 above)
Comment (venturasr) [reply components/knative-function-controller/README.md]: Added!

File: components/knative-function-controller/README.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+- Knative Build (v0.6.0)
+- Knative Serving (v0.6.1)
+- Istio (istio-1.0.7)
+
+## Installation
+
+### Local Deployment
+
+The Manager running locally.
```
[the remark this answers is not in the document]
Comment (venturasr) [reply components/knative-function-controller/README.md]: Yes!

File: components/knative-function-controller/.gitignore [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+# Kubernetes Generated files - skip generated files, except for vendored files
+
+vendor/
+!vendor/**/zz_generated.*
+
+# editor and IDE paraphernalia
+.idea
+*.swp
+*.swo
+*~
```
[the remark this answers is not in the document]
Comment (venturasr) [reply components/knative-function-controller/.gitignore]: I got your point but it is not something that has been discussed in Kyma  so far, or at least I am not aware of. Most of the projects includes their own .gitignore file.

File: components/knative-function-controller/.gitignore (hunk 2 above)
Comment (venturasr) [reply components/knative-function-controller/.gitignore]: That make sense but it is not common in Kyma as every project has their own .gitignore file.

File: components/knative-function-controller/README.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+make install
+```
+
+Run the controller on your machine:
+
+```bash
+make run
+```
+
+#### Manager running inside k8s cluster
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
### Run inside a Kubernetes cluster
```

File: components/knative-function-controller/README.md [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+This workflow needs to be used until kubernetes-sigs/kubebuilder#40 is fixed.
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]: The link to this issue doesn't work. Could you provide me with the valid link? I would include some information on this issue in our doc.
If the issue is fixed, would there be another approach to running the controller on the cluster?

File: components/knative-function-controller/README.md [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+
+```bash
+eval $(minikube docker-env)
+make docker-build
+make install
+make deploy
+```
+
+#### Production Deployment
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
### Run on production
```

File: components/knative-function-controller/README.md [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+Uncomment `manager_image_patch_dev` in `kustomization.yaml`
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
To use the controller on the production environment, uncomment `manager_image_patch_dev` in the `kustomization.yaml` file and run the following:
```

File: components/knative-function-controller/README.md [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+Then run the following commands:
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion

```

File: components/knative-function-controller/README.md [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+make install
+APP_NAME=knative-function-controller
+IMG={DOCKER_PUSH_REPOSITORY}/{DOCKER_PUSH_DIRECTORY}/{APP_NAME}
+TAG={DOCKER_TAG}
+make docker-push
+make deploy
+```
+
+### Test
+
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
To test the controller, run:
```

File: components/knative-function-controller/README.md [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+```bash
+make test
+```
+
+#### Examples
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
## Examples
```

File: components/knative-function-controller/README.md [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+
+Run the examples
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
Run the following examples to learn how to create and manage a function. 
```

File: components/knative-function-controller/README.md [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+
+Create sample function
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
Create a sample function:
```

File: components/knative-function-controller/README.md [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+
+```bash
+kubectl apply -f config/samples/runtime_v1alpha1_function.yaml -n {NAMESPACE}
+```
+
+Search for function
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
Search for a function
```

File: components/knative-function-controller/README.md [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+
+```bash
+kubectl get function -n {NAMESPACE}
+```
+
+```bash
+kubectl get fcn -n {NAMESPACE}
+```
+
+Access the function
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
Access the function:
```

File: components/knative-function-controller/README.md (hunk 11 above)
Comment (venturasr) [reply components/knative-function-controller/README.md]: Done!

File: components/knative-function-controller/Makefile [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+# Deploy controller in the configured Kubernetes cluster in ~/.kube/config
+deploy: manifests install
+	kubectl apply -f config/crds
+	kustomize build config/default | kubectl apply -f -
+	kubectl apply -f config/config.yaml
+
+# Build the docker image
+.PHONY: docker-build
+docker-build: resolve pull-licenses test manifests
+	docker build . -t ${IMG}
```
Comment (antoineco) [inline components/knative-function-controller/Makefile]: It's a make variable so the syntax should be `$(IMG)` (parenthesis) by convention.

File: components/knative-function-controller/Makefile [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+	@echo "updating kustomize image patch file for manager resource"
+	sed -i'' -e 's@image: .*@image: '"${IMG}:${TAG}"'@' ./config/default/manager_image_patch_remote_dev.yaml ./config/default/manager_image_patch_local_dev.yaml
+
+# Push the docker image
+.PHONY: docker-push
+docker-push:
+	docker tag $(APP_NAME) $(IMG):$(TAG)
```
Comment (antoineco) [inline components/knative-function-controller/Makefile]: `APP_NAME` should be IMG`, like in the build step above.

File: components/knative-function-controller/README.md [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+export DOCKER_TAG=<some tag e.g. latest>
+export APP_NAME=knative-function-controller
+export DOCKER_PUSH_REPOSITORY=<e.g. eu.gcr.io or index.docker.io>
+export DOCKER_PUSH_DIRECTORY<e.g. pr or develop>
+make install
+make docker-build
+make docker-push
+make deploy
+```
+
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
### Run on production
```

File: components/knative-function-controller/README.md [hunk 24]
```diff
@@ -0,0 +1,92 @@
+# Knative Function controller
+
+## Overview
+
+The Knative Function controller is a Kubernetes Controller that enables Kyma to manage Function resources.
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
The Knative Function controller is a Kubernetes controller that enables Kyma to manage Function resources.
```

File: components/knative-function-controller/README.md [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+4. Install the CRD to a local Kubernetes cluster:
+
+```bash
+make install
+```
+
+5. Deploy the controller:
+
+```bash
+# Execute this command for local minikube
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
```
I would remove this one.

File: components/knative-function-controller/README.md [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+## Usage
+### Test
+To test the controller, run:
+```bash
+make test
+```
+
+### Examples
+
+Run the following examples to learn how to create and manage a function.
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
Use the following examples to learn how to create and manage a function. 
```

File: components/knative-function-controller/README.md [hunk 27]
```diff
@@ -0,0 +1,92 @@
...
+### Run on production
+To use the controller on the production environment, uncomment the `manager_image_patch_remote_dev.yaml` line  in the `kustomization.yaml` file and follow the instructions for the local installation.
+## Usage
+### Test
+To test the controller, run:
+```bash
+make test
+```
+
+### Examples
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
### Manage functions
```

File: components/knative-function-controller/README.md [hunk 28]
```diff
@@ -0,0 +1,92 @@
...
+export DOCKER_PUSH_DIRECTORY<e.g. pr or develop>
+make install
+make docker-build
+make docker-push
+make deploy
+```
+### Run on production
+To use the controller on the production environment, uncomment the `manager_image_patch_remote_dev.yaml` line  in the `kustomization.yaml` file and follow the instructions for the local installation.
+## Usage
+### Test
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
### Test the controller
```

File: components/knative-function-controller/README.md [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
+```
+
+4. Install the CRD to a local Kubernetes cluster:
+
+```bash
+make install
+```
+
+5. Deploy the controller:
+
```
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
>**NOTE**: Run this command only when you want to deploy the controller locally. It is not necessary for production use.
```

File: components/knative-function-controller/README.md (hunk 25 above)
Comment (bszwarc) [inline components/knative-function-controller/README.md]:
```suggestion
```
I moved it one line up to make an introductory sentence

**Pull request merged**